src/main/scala/org/ergoplatform/modifiers/mempool/ErgoTransaction.scala

```diff
@@ -121,102 +121,103 @@ case class ErgoTransaction(override val inputs: IndexedSeq[Input],
-      // Just to be sure, check that all the input boxes to spend (and to read) are presented.
-      // Normally, this check should always pass, if the client is implemented properly
-      // so it is not part of the protocol really.
-      .validate(txBoxesToSpend, boxesToSpend.size == inputs.size, s"$id: ${boxesToSpend.size} == ${inputs.size}")
-      .validate(txDataBoxes, dataBoxes.size == dataInputs.size, s"$id: ${dataBoxes.size} == ${dataInputs.size}")
-      // Check that there are no overflow in input and output values
-      .validate(txInputsSum, inputSumTry.isSuccess, s"$id")
-      // Check that transaction is not creating money out of thin air.
-      .validate(txErgPreservation, inputSumTry == outputsSumTry, s"$id: $inputSumTry == $outputsSumTry")
-      .validateTry(outAssetsTry, e => ModifierValidator.fatal("Incorrect assets", e)) { case (validation, (outAssets, outAssetsNum)) =>
-        ErgoTransaction.extractAssets(boxesToSpend) match {
-          case Success((inAssets, inAssetsNum)) =>
-            lazy val newAssetId = ByteArrayWrapper(inputs.head.boxId)
-            val tokenAccessCost = stateContext.currentParameters.tokenAccessCost
-            val currentTxCost = validation.result.payload.get
-            // Cost of assets preservation rules checks.
-            // We iterate through all assets to create a map (cost: `(outAssetsNum + inAssetsNum) * tokenAccessCost)`)
-            // and after that we iterate through unique asset ids to check preservation rules (cost: `(inAssets.size + outAssets.size) * tokenAccessCost`)
-            val totalAssetsAccessCost = (outAssetsNum + inAssetsNum) * tokenAccessCost +
-              (inAssets.size + outAssets.size) * tokenAccessCost
-            val newCost = addExact(currentTxCost, totalAssetsAccessCost)
-
-            validation
-              // Check that transaction is not too costly considering all the assets
-              .validate(bsBlockTransactionsCost, maxCost >= newCost, s"$id: assets cost")
-              .validateSeq(outAssets) {
-                case (validationState, (outAssetId, outAmount)) =>
-                  val inAmount: Long = inAssets.getOrElse(outAssetId, -1L)
-
-                  // Check that for each asset output amount is no more than input amount,
-                  // with a possible exception for a new asset created by the transaction
-                  validationState.validate(txAssetsPreservation,
-                    inAmount >= outAmount || (outAssetId == newAssetId && outAmount > 0),
-                    s"$id: Amount in = $inAmount, out = $outAmount. Allowed new asset = $newAssetId, out = $outAssetId")
-              }
-              .payload(newCost)
-          case Failure(e) =>
-            // should never be here as far as we've already checked this when we've created the box
-            ModifierValidator.fatal(e.getMessage)
+    verifier.synchronized {
+      verifier.IR.resetContext() // ensure there is no garbage in the IRContext
```

Comment:
I gave it a second thought and this line kind of invalidate this solution. 
It is impossible to share the same context between two thread, not only becase of race conditions, but
also because this `resetContext` may destroy other thread's intermediate work.
So the solution is to have separate verifiers for each thread. 
It will be faster to have two separate threads working on independent contexts. (no sync, no false sharing, separate hash tables, etc.)

Reply:
Я ещё раз поизучал код и если я гтчегло не упкстил, то едиственным источником гонок является emptyVerifier, который используется в getTxCost в ValidBlocksGenerator.
Таком образом - это всё таки проблема только теста.
И проще всего её исправить, сделав emptyVerifier функцией вместо lazy val.

Reply:
As explained by @catena2w, the script evaluation can be executed from two parallel threads. 
So the problem here is not related to tests. 
The problem is with current production code, which may be not safe.
So, it is necessary to guarantee:
1) that this specific verifier is not used in other methods (i.e. used ONLY in this method)
2) The usage in this method is serialized, i.e. if this method is called in parallel, then invocations are strictly serialized.

Reply:
Script may be evaluated by different threads, but there will be race condition only if them are sharing the same verifier instance. Verifier is passed in validateStateful function as implicit parameter. I have investigated all passes through which this function can be called:
In the following places  new instance of verifier is always created:

* ErgoState.execTransactions
* ErgoMiner.collectTxs
* UtxoStateReader.validateWithCost

The only exception is ValidBlockGenerators.getTxCost which uses ErgoTestConstants.emptyVerifier.

This is why I have made the conclusion that synchronization problems exists only in UtxoTestSpecification parallel test. And it can be fixed by generating new instance of verifier at each access of emptyVerifier.

If you do not agree, please point me at the path where the same verifier instance can be shared between two threads.

Reply:
@knizhnik Well, in this case I don't have reasons to disagree. This was the task, either to fix problem or prove that there is nothing to fix. And your arguments above lead me to the conclusion that there is nothing to fix.
So I'm in favor of accepting this PR.